CMSgov/easi-app
# EASI-1868 Help Send Feedback

# EASI-1868

Design
https://www.figma.com/file/jrc18ENtS3ufj7ttr5vHKW/EASi-Basic-Page-Templates?node-id=28%3A5249

## Changes and Description

This Help / Send Feedback form is a set of text fields and field group selections with an optional additional text field associated to a particular option.

### Highlights
Basic form meta starts with types and constants. Yup features are leveraged as much as possible to drive formik.
- `src/types/helpFeedback.ts`
  Types based on the backend `SendFeedbackEmailInput` type. More types are defined for the ui.
  - `SendFeedbackEmailForm`
    The final form with all ui fields
- `src/constants/helpFeedback.ts`
  - `ADDITIONAL_TEXT_INPUT_SUFFIX`
  - `sendFeedbackOptions`
    Option key value constants for input options. The constant values are sent to the backend and also referenced by translation text.
  - `sendFeedbackOptionFields`
    Option field enums
  - `sendFeedbackOptionFieldsForTextInput`
    Options for associated input text field references
- `src/validations/helpSchema.tsx`
  - `sendFeedbackEmailInputSchema`
    Schema for original backend `SendFeedbackEmailInput`
  - `sendFeedbackEmailFormSchema`
    Ui form schema

The `SendFeedback` form component
- Submit the form with mutation `SendFeedbackEmailQuery`
- `parseForm()` is called during submission to convert the ui `sendFeedbackEmailFormSchema` data to the backend `SendFeedbackEmailInput` data
- "Send and start another form" is handled manually. See the guide https://formik.org/docs/guides/form-submission
- Field components
  - `RadioOptionGroupWithAdditionalText` (general)
  - `CanBeContactedField`
  - `EasiServicesUsedField`

### Notes
- ~~[Yup typescript settings](https://github.com/jquense/yup/blob/pre-v1/docs/typescript.md#typescript-settings) suggests setting `strictNullChecks` and `strictFunctionTypes` but that is held back until further testing across the app. Form type enforcement seems okay as is.~~ 
Already taken care of with `strict: true`
- Formik boolean fields (for yes or no) are treated as strings
  - Their project repo has threads for possible fixes. Might cover that later
- Attempted to implement formik dependent fields for the additional texts inputs, instead of the `parseForm()` stuff, but couldn't get it right. Will find a chance to try again in later work
- Error ui is minimal, with just red bold text. Will cover more of the Truss Field Error Message component later, possibly in the next related help form ticket.
- The separation of types and constants is blurry. Should they all be in one file? Is there a better way to consolidate? Any opinions for better names?
- Would be nice to use translation text in yup schema error message

## How to test this change
Enable the `helpFooter` ld flag

- Jest `src/validations/helpSchema.test.ts`
- Local dev `http://localhost:3000/help/send-feedback`
  Qa the form and see a successful submit or field level errors
  
  Note: There was some jank with the condition behavior between the fields "Would you like your feedback to remain anonymous?" and "May the EASi team contact you for additional information?", but setting `validateOnBlur={false}` & `validateOnChange={false}` seems to hide the problem
  
- Nothing for the form submit mutation yet. Will cover it in the related ticket for the similar "Report a Problem" Help form.
